### PR TITLE
Fixes phpcs  errors and warnings for Magento\Framework\Image

### PR DESCRIPTION
**Pull request by krisdante**
### Description (*)

I was working with those classes and spotted a few warnings and errors
that prevent smooth pull requests. This PR fixes them.
It introduces Localized Exceptions in place of \Exception too.

### Manual testing scenarios (*)

1. `vendor/bin/phpcs --standard=dev/tests/static/framework/Magento/ruleset.xml lib/internal/Magento/Framework/Image/`
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 1]
```diff
@@ -66,7 +71,7 @@ public function backgroundColor($color = null)
      *
      * @param string $filename
      * @return void
-     * @throws \Exception
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:74]: Use imported class name

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 2]
```diff
@@ -77,7 +82,11 @@ public function open($filename)
         try {
             $this->_imageHandler = new \Imagick($this->_fileName);
         } catch (\ImagickException $e) {
-            throw new \Exception(sprintf('Unsupported image format. File: %s', $this->_fileName), $e->getCode(), $e);
+            throw new LocalizedException(
+                new \Magento\Framework\Phrase(sprintf('Unsupported image format. File: %s', $this->_fileName)),
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Whole line can be replaced with `__('Unsupported image format. File: %1', $this->_fileName),`

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 3]
```diff
@@ -86,12 +95,13 @@ public function open($filename)
...
     /**
      * Save image to specific path.
+     *
      * If some folders of path does not exist they will be created
      *
      * @param null|string $destination
      * @param null|string $newName
      * @return void
-     * @throws \Exception  If destination path is not writable
+     * @throws \Magento\Framework\Exception\LocalizedException If destination path is not writable
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:104]: Use imported class name

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 4]
```diff
@@ -241,7 +253,7 @@ public function crop($top = 0, $left = 0, $right = 0, $bottom = 0)
      * @param bool $tile
      * @return void
      * @throws \LogicException
-     * @throws \Exception
+     * @throws \Magento\Framework\Exception\LocalizedException
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:256]: Use imported class name

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 5]
```diff
@@ -333,7 +345,11 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
                 );
             }
         } catch (\ImagickException $e) {
-            throw new \Exception('Unable to create watermark.', $e->getCode(), $e);
+            throw new LocalizedException(
+                new \Magento\Framework\Phrase('Unable to create watermark.'),
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: `__('Unable to create watermark.')`

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 6]
```diff
@@ -346,12 +362,16 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
      * Checks required dependencies
      *
      * @return void
-     * @throws \Exception If some of dependencies are missing
+     * @throws \Magento\Framework\Exception\LocalizedException If some of dependencies are missing
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:328]: Use imported class.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 7]
```diff
@@ -346,12 +362,16 @@ public function watermark($imagePath, $positionX = 0, $positionY = 0, $opacity =
...
      */
     public function checkDependencies()
     {
         if (!class_exists('\Imagick', false)) {
-            throw new \Exception("Required PHP extension 'Imagick' was not loaded.");
+            throw new LocalizedException(
+                new \Magento\Framework\Phrase('Required PHP extension \'Imagick\' was not loaded.'),
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: `__('Required PHP extension "Imagick" was not loaded.')`

File: lib/internal/Magento/Framework/Image/Factory.php [hunk 8]
```diff
@@ -8,6 +8,10 @@
 
 use Magento\Framework\ObjectManagerInterface;
 
+/**
+ * Factory for object to manipulate images
+ *
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Factory.php]: Empty line?

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 9]
```diff
@@ -241,7 +253,7 @@ public function crop($top = 0, $left = 0, $right = 0, $bottom = 0)
      * @param bool $tile
      * @return void
      * @throws \LogicException
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:255]: Import class.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 10]
```diff
@@ -5,6 +5,11 @@
  */
 namespace Magento\Framework\Image\Adapter;
 
+use Magento\Framework\Exception\LocalizedException;
+
+/**
+ * Image adapter from ImageMagick
+ */
 class ImageMagick extends \Magento\Framework\Image\Adapter\AbstractAdapter
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: ~Import class.~

Actually it's in the same namespace, so you can directly use AbstractAdapter without full path.

File: lib/internal/Magento/Framework/Image/Adapter/Config.php [hunk 11]
```diff
@@ -18,13 +18,13 @@ class Config implements ConfigInterface, UploadConfigInterface
...
      * Config path for the maximal image width value
-     * @deprecated
+     * @deprecated Used in a method that is deprecated
      */
     const XML_PATH_MAX_WIDTH_IMAGE = 'system/upload_configuration/max_width';
 
     /**
      * Config path for the maximal image height value
-     * @deprecated
+     * @deprecated Used in a method that is deprecate()
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/Config.php]: What is `deprecate()`?

File: lib/internal/Magento/Framework/Image/Adapter/AbstractAdapter.php [hunk 12]
```diff
@@ -25,24 +25,42 @@ abstract class AbstractAdapter implements AdapterInterface
...
 
+    /**
+     * Put image as tiles next to each other. Used for watermarks.
+     */
     const POSITION_TILE = 'tile';
 
+    /**
+     * Position in the center
+     */
     const POSITION_CENTER = 'center';
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/AbstractAdapter.php]: You may still use single docblock for multiple constants:
```php
    /**#@+
     * Position constants
     */
    public const POSITION_TOP_LEFT = 'top-left';
    ...
    public const POSITION_CENTER = 'center';
    /**#@-*/
```

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Image/Adapter/Config.php (hunk 11 above)
Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/Config.php]: Typo. will fix

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 4 above)
Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:256]: But why? Extracted documentation is not aware of the imports. 
I see those everywhere in Magento Code:
```
    /**
     * Remove product from Bundle product option
     *
     * @param string $sku
     * @param int $optionId
     * @param string $childSku
     * @throws \Magento\Framework\Exception\NoSuchEntityException
     * @throws \Magento\Framework\Exception\InputException
     * @return bool
     */
    public function removeChild($sku, $optionId, $childSku);
```

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 5 above)
Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: I think the whole idea behind LocalizedException and Phrase is not to use translation function on top of it.

Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Nowhere in code it is used this way

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 1 above)
Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:74]: Let's discuss this in one place

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 5 above)
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: `__()` actually is new Phrase - take a look into https://github.com/magento/magento2/blob/c7c0a992d83de56f15555d3edcecef7aaa32f0d7/app/functions.php#L19:29

Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: So using new Phrase is cleaner (OOP not superglobal function)

Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Both ways are correct unless you use full path for all `Phrase`s :-)

Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: OK i checked other places in Magento. Changed

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 4 above)
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:256]: Let me discuss that. Let's keep it like this for now :-)

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 6 above)
Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:328]: Mentioned previously.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 13]
```diff
@@ -499,4 +480,84 @@ protected function _getImagickPixelObject($color = null)
...
+            $offsetY += $watermark->getImageHeight();
+        }
+    }
+
+    /**
+     * Add watermark at given X,Y position
+     *
+     * @param $positionX
+     * @param $positionY
+     * @param \Imagick $watermark
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:552]: I'd import the `\Imagick` class.

**Comment by lbajsarowicz** [comment]
@krisdante Fix the Static tests and import `Imagick` class, then I'll approve PR :+1:

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 13 above)
Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:552]: What is the benefit of such import?

Comment (lbajsarowicz) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:552]: The imports are consistent.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 14]
```diff
@@ -499,4 +480,84 @@ protected function _getImagickPixelObject($color = null)
     {
         return new \ImagickPixel($color);
     }
+
+    /**
+     * Resizes watermark to desired size, when it is not stretched
+     *
+     * @param \Imagick $watermark
+     */
+    private function resizeWatermerk(\Imagick $watermark): void
```
Comment (rodde177) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: typo > `resizeWatermerk`

**Comment by lbajsarowicz** [comment]
@magento run tests

**Comment by krisdante** [comment]
Jezus this is boring.... what now. Why the way to test is locally is not documented. phpcs is passing for me.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 13 above)
Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:552]: I followed this suggestion and phpstan raged. Reversing

**Comment by krisdante** [comment]
```
/usr/bin/php /var/www/magentosrc/./vendor/phpunit/phpunit/phpunit -v --testdox testsuite/Magento/Test/Php/LiveCodeTest.php
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.27 with Xdebug 2.7.2
Configuration: /var/www/creativeshop/magentosrc/dev/tests/static/phpunit.xml.dist

Magento\Test\Php\LiveCode
 [x] Code style
 [x] Code mess
 [x] Copy paste
 [x] Strict types
 [x] Php compatibility
 [x] Php stan
```

Those tests are passing with my clone of Magento. I do not know why they are failing for B2B test here. I cannot reproduce it. Also, the failed functional test has nothing to do with the change here .

**Comment by lbajsarowicz** [comment]
@krisdante Don't worry about Functional Tests - this is known bug of Magento `2.4-develop` branch and will be merged soon to `2.4-develop` (then you'll be able to rebase your branch onto it). Unless - just fix static tests.or optionally restart them if you believe that your code is fine.

Use
```
@magento run Static Tests
```

**Comment by krisdante** [comment]
I do not know how to fix them. They are passing on local. See my message.

**Comment by lbajsarowicz** [comment]
@krisdante Let me ping the right person for that kind of issue.
Looks like a missing library in the Testing environment...

**Comment by lenaorobei** [comment]
This issue is caused because `ImageMagick` adapter is non default one and can be configured. 

Please add `lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php` to the `dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/blacklist/common.txt`.

**Comment by lbajsarowicz** [comment]
@krisdante Could you make the changes requested by @lenaorobei, please?

**Comment by krisdante** [comment]
@lbajsarowicz I did. Static tests are now passing. Failing B2B test is irrelevant.

**Review by lbajsarowicz — APPROVED** [review]
Thank you for the great contribution!

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7260](https://jira.corp.magento.com/browse/ENGCOM-7260) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by krisdante** [comment]
This DB error is false. I am not going to re-run this test anymore. Merge it or delete the whole PR.

**Comment by lbajsarowicz** [comment]
@krisdante Don't worry about false negatives. These are also verified manually by Commeng team, so as long as it's false negative, we're able to skip it.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

vendor/bin/phpcs --standard=dev/tests/static/framework/Magento/ruleset.xml lib/internal/Magento/Framework/Image/

**Before:**
Screenshot from 2020-04-07 14-27-30

**After:**
phpcs

**Comment by engcom-Echo** [comment]
Failed Database Compare not related to the changes in this PR

**Pull request merged**